### PR TITLE
XDR mirroring

### PR DESCRIPTION
**Pull request by roysagi**
## Status
- [x] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/25898

## Description
Adding XDR incident mirroring:

- [x] make sure which fields return via the API and to which fields should they be mapped.
- [x] in fetch incidents -> list all incidents -> get extra data for all incident ID
- [x] create get-remote-data 
- [x] create update-remote-system
- [x] create get-mapping fields
- [x] create mapping from XDR severity to Demisto severity (existing script xdr sync already does it)
- [x] map xdr status to incident status and upon incident closure in demisto ask to provide reason.
- [ ] check incident handling v3 error
- [x] deprecate xdr sync script from version 6.0.0
- [x] change xdrstatus to be a single select field with "under_investigation" and "new" as options.
- [x] Add checkbox param to sync incident owners (uses getUserByUsername and getUserByMail methods)
- [x] Modify alert fields to be without "_" - make sure all significant alerts fields are mapped to the grid layout.
- [x] Add sync parameter for dbotMirrorDirection (incoming, outgoing, Both, None(default)), needs to have clear tooltip
- [x] Add mirroring to dbotMirrorInstance.
- [x] Add error entry in cases of failed incoming and outgoing commands
   - [x] force error experiment - empty error is being printed, checking cause
- [x] Add last sync time to layout
- [x] Create new field - sync error

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [ ] 4.5.0
- [ ] 5.0.0
- [ ] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation 

## Demisto Partner?
- [ ] The title must be in the following format: **[YOUR_PARTNER_ID] short description**

### CODE REVIEW

File: Packs/CortexXDR/pack_metadata.json [hunk 1]
```diff
@@ -2,7 +2,7 @@
     "name": "Palo Alto Networks Cortex XDR - Investigation and Response",
     "description": "This Content Pack automates Cortex XDR incident response, and includes custom Cortex XDR incident views and layouts to aid analyst investigations.",
     "support": "xsoar",
-    "currentVersion": "2.2.3",
+    "currentVersion": "2.2.4",
```
Comment (ronykoz) [inline Packs/CortexXDR/pack_metadata.json:5]: this should be a minor, not revision

File: Packs/CortexXDR/Playbooks/PaloAltoNetworks_Cortex_XDR_Incident_Sync.yml [hunk 2]
```diff
@@ -230,5 +230,7 @@ view: |-
   }
 inputs: []
 outputs: []
+fromvesion: 6.0.0
```
Comment (ronykoz) [inline Packs/CortexXDR/Playbooks/PaloAltoNetworks_Cortex_XDR_Incident_Sync.yml:233]: you have another fromversion at the top, remove that one

File: Packs/CortexXDR/Integrations/PaloAltoNetworks_XDR/PaloAltoNetworks_XDR.yml [hunk 3]
```diff
@@ -1589,6 +1589,30 @@ script:
       run this command with no filters.
     execution: true
     name: xdr-endpoint-scan
+  - deprecated: false
+    description: Get mapping fields from remote incident.
```
Comment (ronykoz) [inline Packs/CortexXDR/Integrations/PaloAltoNetworks_XDR/PaloAltoNetworks_XDR.yml:1593]: add a disclaimer that this is mainly for debugging purposes

File: Packs/CortexXDR/Playbooks/Cortex_XDR_incident_handling_v3.yml [hunk 4]
```diff
@@ -0,0 +1,1776 @@
+id: Cortex XDR incident handling v3
```
Comment (ronykoz) [inline Packs/CortexXDR/Playbooks/Cortex_XDR_incident_handling_v3.yml:1]: let's review it in the demo

File: Packs/CortexXDR/Integrations/PaloAltoNetworks_XDR/PaloAltoNetworks_XDR.py [hunk 5]
```diff
@@ -1600,6 +1637,139 @@ def endpoint_scan_command(client, args):
...
+        xsoar_field_name = XDR_INCIDENT_FIELDS[field].get('xsoar_field_name')
+        if xsoar_field_name in delta:
+            if xsoar_field_name == 'severity':
+                severity_mapping = {
+                    1: "low",
+                    2: "medium",
+                    3: "high",
+                    4: "high"
+                }
+                update_args['manual_severity'] = severity_mapping[delta.get('severity')]
```
Comment (ronykoz) [inline Packs/CortexXDR/Integrations/PaloAltoNetworks_XDR/PaloAltoNetworks_XDR.py:1734]: why not to make it a part of the mapper

File: Packs/CortexXDR/Integrations/PaloAltoNetworks_XDR/PaloAltoNetworks_XDR.py [hunk 6]
```diff
@@ -1614,12 +1784,35 @@ def fetch_incidents(client, first_fetch_time, last_run: dict = None):
...
+            'xdrmediumseverityalertcount': incident_data.get('med_severity_alert_count'),
+            'xdrhighseverityalertcount': incident_data.get('high_severity_alert_count'),
+            'xdrusercount': incident_data.get('user_count'),
+            'xdrhostcount': incident_data.get('host_count'),
+            'xdrnotes': incident_data.get('notes'),
+            'xdrurl': incident_data.get('xdr_url'),
+            'xdralerts': incident_data.get('alerts'),
+            'xdrresolvecomment': incident_data.get('resolve_comment'),
+            'xdrnetworkartifacts': incident_data.get('network_artifacts'),
+            'xdrfileartifacts': incident_data.get('file_artifacts'),
```
Comment (ronykoz) [inline Packs/CortexXDR/Integrations/PaloAltoNetworks_XDR/PaloAltoNetworks_XDR.py:1815]: don't we want all of those in the mapping?

Comment (anara123) [reply Packs/CortexXDR/Integrations/PaloAltoNetworks_XDR/PaloAltoNetworks_XDR.py:1815]: correct. it should be in the mapping

File: Packs/CortexXDR/pack_metadata.json [hunk 7]
```diff
@@ -2,7 +2,7 @@
     "name": "Palo Alto Networks Cortex XDR - Investigation and Response",
     "description": "This Content Pack automates Cortex XDR incident response, and includes custom Cortex XDR incident views and layouts to aid analyst investigations.",
     "support": "xsoar",
-    "currentVersion": "2.2.4",
+    "currentVersion": "2.4.0",
```
Comment (ronykoz) [inline Packs/CortexXDR/pack_metadata.json:5]: shouldn't it be 2.3.0?

File: Packs/CortexXDR/Scripts/XDRSyncScript_5_9_9/CHANGELOG.md [hunk 8]
```diff
@@ -0,0 +1,8 @@
+## [Unreleased]
+
+
+## [20.2.4] - 2020-02-25
+Fixed an issue where incident in Demisto wasn't updated if there was a change in XDR
+
+## [20.2.0] - 2020-02-04
+- Updated outputs and added additional alert outputs.
```
Comment (ronykoz) [inline Packs/CortexXDR/Scripts/XDRSyncScript_5_9_9/CHANGELOG.md:8]: remove, no need

File: Packs/CortexXDR/Playbooks/PaloAltoNetworks_Cortex_XDR_Incident_Sync.yml [hunk 9]
```diff
@@ -230,7 +230,5 @@ view: |-
   }
 inputs: []
```
Comment (ronykoz) [inline Packs/CortexXDR/Playbooks/PaloAltoNetworks_Cortex_XDR_Incident_Sync.yml]: do we need this from 6.0?

File: Packs/CortexXDR/Scripts/XDRSyncScript_5_9_9/CHANGELOG.md (hunk 8 above)
Comment (roysagi) [reply Packs/CortexXDR/Scripts/XDRSyncScript_5_9_9/CHANGELOG.md:8]: ok

File: Packs/CortexXDR/pack_metadata.json (hunk 7 above)
Comment (roysagi) [reply Packs/CortexXDR/pack_metadata.json:5]: it is 2.3.0 right now

File: Packs/CortexXDR/Playbooks/PaloAltoNetworks_Cortex_XDR_Incident_Sync.yml (hunk 9 above)
Comment (roysagi) [reply Packs/CortexXDR/Playbooks/PaloAltoNetworks_Cortex_XDR_Incident_Sync.yml]: Good question - I think not - @anara123?

Comment (anara123) [reply Packs/CortexXDR/Playbooks/PaloAltoNetworks_Cortex_XDR_Incident_Sync.yml]: We don't

File: Packs/CortexXDR/Integrations/PaloAltoNetworks_XDR/PaloAltoNetworks_XDR.py [hunk 10]
```diff
@@ -1638,12 +1794,17 @@ def fetch_incidents(client, first_fetch_time, last_run: dict = None):
 
     for raw_incident in raw_incidents:
         incident_id = raw_incident.get('incident_id')
+        incident_data = get_incident_extra_data_command(client, {"incident_id": incident_id,
```
Comment (anara123) [inline Packs/CortexXDR/Integrations/PaloAltoNetworks_XDR/PaloAltoNetworks_XDR.py:1797]: lets add a new boolean parameter `fetch incident alerts and artifacts` with default value True, and if this param enabled, call get extra data.

Comment (roysagi) [reply Packs/CortexXDR/Integrations/PaloAltoNetworks_XDR/PaloAltoNetworks_XDR.py:1797]: done

File: Packs/CortexXDR/Integrations/PaloAltoNetworks_XDR/PaloAltoNetworks_XDR.py [hunk 11]
```diff
@@ -1729,55 +1746,215 @@ def get_remote_data_command(client, args):
...
-            mirrored_object=incident_data,
-            entries=[closing_entry] if closing_entry else []
-        )
+    return closing_entry
 
-    else:
-        # no new data modified so not updating the incident
-        return {}
 
+def handle_incoming_error_in_mirror(incident_data, error):
```
Comment (ronykoz) [inline Packs/CortexXDR/Integrations/PaloAltoNetworks_XDR/PaloAltoNetworks_XDR.py:1752]: let's add this one in commonserverpython

**Comment by roysagi** [comment]
Closing as this got corrupted.
Continuation in: https://github.com/demisto/content/pull/8526

**Pull request closed without merging**